File: generate/reference/duality-of-compilation/ocaml/sum_range.ml
let rec range i n =
  if i < n then
    i :: range (i + 1) n
  else
    []

let rec sum xs =
  match xs with
  | [] -> 0
  | y :: ys -> y + sum ys

let main n =
  sum (range 0 n)

let () =
  match Sys.argv with
  | [| _ |] -> print_endline (string_of_int (main 10))
  | [| _; s |] -> (
      match int_of_string_opt s with
      | Some n -> print_endline (string_of_int (main n))
      | None -> print_endline "argument is not a number")
  | _ -> print_endline "arguments are too many"

File: generate/reference/duality-of-compilation/rust/sum_range.rs
use std::env;
use std::rc::Rc;

enum List {
    Nil,
    Cons(isize, Rc<List>),
}

fn range(i: isize, n: isize) -> List {
    if i < n {
        List::Cons(i, Rc::new(range(i + 1, n)))
    } else {
        List::Nil
    }
}

fn sum(list: &List) -> isize {
    match list {
        List::Nil => 0,
        List::Cons(y, ys) => y + sum(ys),
    }
}

fn run(n: isize) -> isize {
    sum(&range(0, n))
}

fn main() {
    let args: Vec<String> = env::args().collect();

    let arg = match args.get(1) {
        None => 10,
        Some(val) => match val.parse::<isize>() {
            Ok(val) => val,
            Err(_) => {
                println!("argument is not a number");
                return;
            }
        },
    };

    println!("{}", run(arg));
}
